deep_qa/data/vocabulary.py

```diff
@@ -0,0 +1,119 @@
+        self._finalized = False
+
+    def set_from_file(self, filename: str, oov_token: str="@@UNKNOWN@@", namespace: str="tokens"):
+        self._oov_token = oov_token
+        self._token_indices[namespace] = {self._padding_token: 0}
+        self._reverse_token_indices[namespace] = {0: self._padding_token}
+        with codecs.open(filename, 'r', 'utf-8') as input_file:
+            for i, line in enumerate(input_file.readlines()):
+                token = line[:-1]  # remove the newline
+                self._token_indices[namespace][token] = i + 1
+                self._reverse_token_indices[namespace][i + 1] = token
+
+    def finalize(self):
+        logger.info("Finalizing data indexer")
+        self._finalized = True
+
+    def fit_token_dictionary(self, dataset, min_count: int=1):
+        """
+        Given a ``Dataset``, this method decides which tokens are given an index, and which ones are
+        mapped to an OOV token (in this case "UNK").  This method must be called before any dataset
+        is indexed with this ``Vocabulary``.  If you don't first fit the token dictionary, you'll
+        basically map every token onto "UNK".
+
+        We call ``instance.words()`` for each instance in the dataset, and then keep all tokens that
+        appear at least ``min_count`` times.
+
+        Parameters
+        ----------
+        dataset: ``TextDataset``
+            The dataset to index.
+
+        min_count: int, optional (default=1)
+            The minimum number of occurences a token must have in the dataset in order to be
+            assigned an index.
+        """
+        # TODO(matt): FIXME
+        logger.info("Fitting token dictionary with min count of %d, finalized is %s",
+                    min_count, self._finalized)
+        if self._finalized:
+            logger.warning("Trying to fit a finalized Vocabulary.  This is a no-op.  Did you "
+                           "really want to do this?")
+            return
+        namespace_token_counts = defaultdict(lambda: defaultdict(int))
+        for instance in tqdm.tqdm(dataset.instances):
+            namespace_dict = instance.words()
+            for namespace in namespace_dict:
+                for token in namespace_dict[namespace]:
+                    namespace_token_counts[namespace][token] += 1
+        for namespace in tqdm.tqdm(namespace_token_counts):
+            for token, count in namespace_token_counts[namespace].items():
+                if count >= min_count:
+                    self.add_token_to_index(token, namespace)
+
+    def increment_token_count(self, token: str, namespace: str='tokens', count: int=1):
+        """
+        Increments the count of ``token`` in ``namespace`` by ``count``.  This is used for
+        determining which tokens are included in the vocabulary.
+        """
+        self._token_counts[namespace][token] += count
+
+    def add_token_to_index(self, token: str, namespace: str='tokens') -> int:
```

Comment:
Can we call this and the method below `add_token_to_namespace` rather than `index`. Everyone will know what the index of a `Vocabulary` object does, but this is specifically adding a token to a particular namespace.